src/main/java/de/tum/in/www1/artemis/versioning/VersionRangesRequestCondition.java

```diff
@@ -0,0 +1,342 @@
+            }
+            if (!combined) {
+                newRanges.add(selectedRange);
+            }
+        }
+        return newRanges;
+    }
+
+    /**
+     * Checks if the given condition collides with this condition.
+     *
+     * @param versionRangesRequestCondition The condition to check.
+     * @return True if the conditions collide, false otherwise.
+     */
+    public boolean collide(@NotNull VersionRangesRequestCondition versionRangesRequestCondition) {
+        Set<VersionRange> tbcRanges = versionRangesRequestCondition.getRanges();
+
+        // If one of those represents all versions, the other one has to collide
+        if (ranges.isEmpty() || tbcRanges.isEmpty()) {
+            return true;
+        }
+
+        for (VersionRange range : ranges) {
+            for (VersionRange tbcRange : tbcRanges) {
+                if (inRangeCodes.contains(VersionRangeComparator.compare(range, tbcRange))) {
+                    return true;
+                }
+            }
+        }
+        return false;
+    }
+
+    public Set<VersionRange> getRanges() {
+        return ranges;
+    }
+
+    @Override
+    public int compareTo(@NotNull VersionRangesRequestCondition versionRangesRequestCondition, @NotNull HttpServletRequest httpServletRequest) {
+        if (this.getMatchingCondition(httpServletRequest) == this) {
+            return -1;
+        }
+        else if (versionRangesRequestCondition.getMatchingCondition(httpServletRequest) == versionRangesRequestCondition) {
+            return 1;
+        }
+        else {
+            return 0;
+        }
+    }
+
+    /**
+     * Returns the condition that matches the request or {@code null} if there is no match.
+     *
+     * @param request the current request
+     * @return the matching condition or {@code null}
+     */
+    @Nullable
+    @Override
+    public VersionRangesRequestCondition getMatchingCondition(@NotNull HttpServletRequest request) {
+        String path = request.getRequestURI();
+        if (!path.matches("^/?api/.*")) {
+            return null;
+        }
+        // If any version is valid, return this
+        if (ranges.isEmpty()) {
+            return this;
+        }
+
+        for (VersionRange range : ranges) {
+            if (versionRangeMatchesRequest(range, request)) {
+                return this;
+            }
+        }
+
+        return null;
+    }
+
+    /**
+     * Checks if the range matches the request.
+     *
+     * @param range the version range to match
+     * @param request the current request
+     * @return true if the range matches the request, false otherwise
+     */
+    public boolean versionRangeMatchesRequest(@NotNull VersionRange range, @NotNull HttpServletRequest request) {
+        String path = request.getRequestURI();
+        // Expect optional leading slash, "api/", optional numerical version with a "v" prefix and some path afterwards
+        var pattern = Pattern.compile("^/?api(/v(\\d*))?/.+$");
```

Comment:
Saving the compiled pattern into a constant would avoid recompiling multiple times for every request.